extensions/ql-vscode/src/skeleton-query.ts

```diff
@@ -0,0 +1,218 @@
+import { join } from "path";
+import { CancellationToken, Uri, workspace } from "vscode";
+import { CodeQLCliServer } from "./cli";
+import { OutputChannelLogger } from "./common";
+import { Credentials } from "./common/authentication";
+import { QueryLanguage } from "./common/query-language";
+import { askForLanguage, isFolderAlreadyInWorkspace } from "./helpers";
+import { getErrorMessage } from "./pure/helpers-pure";
+import { QlPackGenerator } from "./qlpack-generator";
+import { DatabaseItem, DatabaseManager } from "./local-databases";
+import * as databaseFetcher from "./databaseFetcher";
+import { ProgressCallback } from "./progress";
+
+type QueryLanguagesToDatabaseMap = {
+  [id: string]: string;
+};
+
+export class SkeletonQueryWizard {
+  private language: string | undefined;
+  private folderName: string | undefined;
+  private fileName = "example.ql";
+
+  QUERY_LANGUAGE_TO_DATABASE_REPO: QueryLanguagesToDatabaseMap = {
+    csharp: "github/codeql",
+    python: "github/codeql",
+    ruby: "github/codeql",
+    javascript: "github/codeql",
+    go: "github/codeql",
+    ql: "github/codeql",
+  };
+
+  constructor(
+    private readonly cliServer: CodeQLCliServer,
+    private readonly storagePath: string | undefined,
+    private readonly progress: ProgressCallback,
+    private readonly credentials: Credentials | undefined,
+    private readonly extLogger: OutputChannelLogger,
+    private readonly databaseManager: DatabaseManager,
+    private readonly token: CancellationToken,
+  ) {}
+
+  public async execute() {
+    // show quick pick to choose language
+    await this.chooseLanguage();
+    if (!this.language) {
+      return;
+    }
+
+    this.folderName = `codeql-custom-queries-${this.language}`;
+    const skeletonPackAlreadyExists = isFolderAlreadyInWorkspace(
+      this.folderName,
+    );
+
+    if (skeletonPackAlreadyExists) {
+      // just create a new example query file in skeleton QL pack
+      await this.createExampleFile();
+      // select existing database for language
+      await this.selectExistingDatabase();
+    } else {
+      // generate a new skeleton QL pack with query file
+      await this.createQlPack();
+      // download database based on language and select it
+      await this.downloadDatabase();
+    }
+
+    // open a query file
+    await this.openExampleFile();
+  }
+
+  private async openExampleFile() {
+    if (this.folderName === undefined || this.storagePath === undefined) {
```

Comment:
I'm not really sure it would make it more readable. We'd still need this check to see if they're undefined, when we call the `openExampleFile` method. But now we also have to pass in params to the method. 

We'd be inconsistent if we didn't also do this for `this.fileName` and also for the previous methods that use these values, so I think it might make it harder to read because we'd be getting some values from state and some from params. 

I might lack imagination, I admit. I think it would result in longer method calls.